<a href="https://colab.research.google.com/github/TimothyAv/Ai/blob/main/12204574_week6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

***Deep Learning***
====

In [ ]:
import tensorflow as tf

# Load the MNIST dataset
mnist = tf.keras.datasets.mnist

# Split the dataset into training and testing sets
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Normalize the pixel values to the range [0, 1]
train_images = train_images / 255.0
test_images = test_images / 255.0

11490434/11490434 [==============================] - 0s 0us/step


The code above loads the MNIST dataset, which contains handwritten digits,and splits it into training and testing sets. It then normalizes the pixel values to ensure they are in the range [0, 1], which is a common preprocessing step when working with image data in neural networks.

*Activity 2.1*
==

In [ ]:
model = tf.keras.Sequential([
    tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D(64, (3, 3), activation='relu'),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dropout(0.5),  # Add Dropout layer with a dropout rate
    tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(train_images, train_labels, epochs=5)

test_loss, test_acc = model.evaluate(test_images, test_labels)
model.save('mnist_cnn_model_with_dropout_bn.h5')

loaded_model = tf.keras.models.load_model('mnist_cnn_model_with_dropout_bn.h5')

test_loss, test_acc = loaded_model.evaluate(test_images, test_labels)

print("Test accuracy:", test_acc)

Epoch 1/5
1875/1875 [==============================] - 41s 21ms/step - loss: 0.2006 - accuracy: 0.9405
Epoch 2/5
1875/1875 [==============================] - 39s 21ms/step - loss: 0.0962 - accuracy: 0.9736
Epoch 3/5
1875/1875 [==============================] - 40s 21ms/step - loss: 0.0703 - accuracy: 0.9806
Epoch 4/5
1875/1875 [==============================] - 39s 21ms/step - loss: 0.0607 - accuracy: 0.9829
Epoch 5/5
313/313 [==============================] - 2s 6ms/step - loss: 0.0353 - accuracy: 0.9907


/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3000: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


313/313 [==============================] - 2s 7ms/step - loss: 0.0353 - accuracy: 0.9907
Test accuracy: 0.9907000064849854


This program constructs a batch normalization and dropout layer convolutional neural network (CNN) model, trains it on the MNIST dataset for digit recognition, and assesses its performance. To do this, a network architecture is specified, it is compiled with an optimizer and a loss function, it is trained for five epochs, and finally the trained model is saved and loaded to test its accuracy on a test dataset.

***Test accuracy:  0.9907000064849854***

*Activity 2.2*
==

In [ ]:
# Define a list of activation functions to experiment with: This list contains different activation functions, such as sigmoid, tanh, and relu, which will be used to modify the activation functions in the Convolutional Neural Network (CNN) layers for experimentation.
activation_functions = ['sigmoid', 'tanh', 'relu']

for activation_function in activation_functions:
    # Build the CNN model with the specified activation function: A CNN model is constructed, and the specified activation function is applied to its convolutional and dense layers.
    model = tf.keras.Sequential([
        tf.keras.layers.Conv2D(32, (3, 3), activation=activation_function, input_shape=(28, 28, 1)),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Conv2D(64, (3, 3), activation=activation_function),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(128, activation=activation_function),
        tf.keras.layers.Dense(10, activation='softmax')
    ])

    # Compile the model: The model is configured with the Adam optimizer and sparse categorical cross-entropy loss for training, as well as accuracy as a metric for evaluation.

    model.compile(optimizer='adam',
                  loss='sparse_categorical_crossentropy',
                  metrics=['accuracy'])

    # Train the model: The CNN model is trained using the training images and labels for a fixed number of epochs (5 in this case).

    model.fit(train_images, train_labels, epochs=5)

    # Evaluate the model: The performance of the trained model is assessed using the test dataset, and its accuracy is computed.

    test_loss, test_acc = model.evaluate(test_images, test_labels)

    print(f"Activation Function: {activation_function}")  # Display the activation function being used in this iteration.
    print("Test accuracy:", test_acc)  # Display the test accuracy achieved by the model.
    print("------------------")  # Separating lines for clarity between different activation function experiments.


Epoch 1/5
1875/1875 [==============================] - 34s 18ms/step - loss: 0.5673 - accuracy: 0.8151
Epoch 2/5
1875/1875 [==============================] - 32s 17ms/step - loss: 0.1094 - accuracy: 0.9675
Epoch 3/5
1875/1875 [==============================] - 33s 18ms/step - loss: 0.0755 - accuracy: 0.9761
Epoch 4/5
1875/1875 [==============================] - 33s 17ms/step - loss: 0.0576 - accuracy: 0.9827
Epoch 5/5
313/313 [==============================] - 2s 6ms/step - loss: 0.0470 - accuracy: 0.9845
Activation Function: sigmoid
Test accuracy: 0.984499990940094
------------------
Epoch 1/5
1875/1875 [==============================] - 33s 17ms/step - loss: 0.1237 - accuracy: 0.9619
Epoch 2/5
1875/1875 [==============================] - 33s 18ms/step - loss: 0.0452 - accuracy: 0.9863
Epoch 3/5
1875/1875 [==============================] - 33s 18ms/step - loss: 0.0301 - accuracy: 0.9910
Epoch 4/5
1875/1875 [==============================] - 32s 17ms/step - loss: 0.0219 - accuracy: 0.9

**Activation functions:**

**sigmoid - test accuracy: 0.984499990940094**

**tanh - test accuracy: 0.9879000186920166**

**relu - test accuracy 0.9883999824523926**



This program defines a list of activation functions (sigmoid, tanh, and relu) for use in convolutional neural network (CNN) experiments. A CNN model is built, trained, and tested using the provided activation function on convolutional and dense layers for each activation function, with the results for each activation function being shown.


*Activity 2.3*
==

In [5]:
model = tf.keras.Sequential([
    tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D(64, (3, 3), activation='relu'),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D(128, (3, 3), activation='relu'),  # Additional Conv2D layer
    tf.keras.layers.MaxPooling2D((2, 2)),  # Additional MaxPooling2D layer
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(train_images, train_labels, epochs=5)

test_loss, test_acc = model.evaluate(test_images, test_labels)
print("Test accuracy:", test_acc)

Epoch 1/5
1875/1875 [==============================] - 35s 18ms/step - loss: 0.1952 - accuracy: 0.9388
Epoch 2/5
1875/1875 [==============================] - 33s 18ms/step - loss: 0.0687 - accuracy: 0.9791
Epoch 3/5
1875/1875 [==============================] - 32s 17ms/step - loss: 0.0483 - accuracy: 0.9854
Epoch 4/5
1875/1875 [==============================] - 33s 17ms/step - loss: 0.0362 - accuracy: 0.9887
Epoch 5/5
313/313 [==============================] - 2s 6ms/step - loss: 0.0468 - accuracy: 0.9880
Test accuracy: 0.9879999756813049


Convolutional neural network (CNN) model for image categorization is defined by this code. The model is trained on training pictures using the Adam optimizer and assessed on a test dataset to calculate and show its test accuracy. The model is composed of three convolutional layers with increasing filter sizes and max-pooling layers, followed by two fully connected layers.

*Activity 2.4*
==

In [6]:

optimizers = ['sgd', 'rmsprop']

for optimizer in optimizers:
    model = tf.keras.Sequential([
        tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(128, activation='relu'),
        tf.keras.layers.Dense(10, activation='softmax')
    ])

    model.compile(optimizer=optimizer, loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    model.fit(train_images, train_labels, epochs=5)

    test_loss, test_acc = model.evaluate(test_images, test_labels)
    print(f"Test accuracy with {optimizer} optimizer:", test_acc)

Epoch 1/5
1875/1875 [==============================] - 21s 11ms/step - loss: 0.4583 - accuracy: 0.8741
Epoch 2/5
1875/1875 [==============================] - 21s 11ms/step - loss: 0.2167 - accuracy: 0.9365
Epoch 3/5
1875/1875 [==============================] - 20s 11ms/step - loss: 0.1609 - accuracy: 0.9531
Epoch 4/5
1875/1875 [==============================] - 20s 10ms/step - loss: 0.1293 - accuracy: 0.9619
Epoch 5/5
313/313 [==============================] - 1s 4ms/step - loss: 0.0942 - accuracy: 0.9724
Test accuracy with sgd optimizer: 0.9724000096321106
Epoch 1/5
1875/1875 [==============================] - 23s 12ms/step - loss: 0.1550 - accuracy: 0.9529
Epoch 2/5
1875/1875 [==============================] - 22s 12ms/step - loss: 0.0519 - accuracy: 0.9844
Epoch 3/5
1875/1875 [==============================] - 23s 12ms/step - loss: 0.0356 - accuracy: 0.9893
Epoch 4/5
1875/1875 [==============================] - 22s 12ms/step - loss: 0.0250 - accuracy: 0.9922
Epoch 5/5
313/313 [=====

Using two alternative optimizers, SGD (Stochastic Gradient Descent) and RMSprop, a convolutional neural network (CNN) model for image classification is trained in this code. Convolutional and dense layers are included in the model, which is trained using each optimizer for 5 epochs. The test accuracy is assessed and shown for each optimizer independently.

**Test accuracy with sgd optimizer: 0.9724000096321106**

**Test accuracy with rmsprop optimizer: 0.9865999817848206**

*Activity 2.5*
==

In [7]:
model = tf . keras . Sequential ([
    tf.keras.layers.Conv2D(32 , (3,3) , activation ='relu',input_shape =(28 , 28 , 1) ) ,
    tf.keras.layers.MaxPooling2D ((2 , 2) ) ,
    tf.keras.layers.Conv2D(64 , (3 ,3) , activation ='relu') ,
    tf.keras.layers.MaxPooling2D((2 , 2) ) ,
    tf.keras.layers.Flatten() ,
    tf.keras.layers.Dense(128 , activation ='relu') ,
    tf.keras.layers.Dense(10 , activation ='softmax')
])


model.compile( optimizer ='adam' ,loss = 'sparse_categorical_crossentropy',metrics = [ 'accuracy'])
model.fit( train_images , train_labels , epochs =5)

test_loss , test_acc = model . evaluate ( test_images , test_labels)
model.save ('mnist_cnn_model.h5')

loaded_model = tf.keras.models.load_model ('mnist_cnn_model.h5')

test_loss , test_acc = loaded_model.evaluate (test_images ,test_labels)

print("Test accuracy:", test_acc)

Epoch 1/5
1875/1875 [==============================] - 31s 16ms/step - loss: 0.1262 - accuracy: 0.9618
Epoch 2/5
1875/1875 [==============================] - 31s 16ms/step - loss: 0.0431 - accuracy: 0.9865
Epoch 3/5
1875/1875 [==============================] - 31s 17ms/step - loss: 0.0291 - accuracy: 0.9908
Epoch 4/5
1875/1875 [==============================] - 30s 16ms/step - loss: 0.0206 - accuracy: 0.9932
Epoch 5/5
313/313 [==============================] - 2s 7ms/step - loss: 0.0397 - accuracy: 0.9876
Test accuracy: 0.9876000285148621


This code defines a convolutional neural network (CNN) model for image classification that consists of fully connected layers, max-pooling layers, and convolutional layers using ReLU activation functions. The Adam optimizer and sparse categorical cross-entropy loss are used to build the model, which is then trained for five iterations on a dataset of training pictures and labels.

The model is stored to a file called "mnist_cnn_model.h5" after training. It is then loaded back into memory with the use of the 'tf.keras.models.load_model' command, and its performance is assessed using a test dataset, with the test accuracy being output on the terminal. This shows how trained models may be saved and loaded for use at a later time.

**Test accuracy: 0.9876000285148621**